sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java

```diff
@@ -120,4 +130,115 @@ public void processElement() {}
+
+              @ProcessElement
+              public void foo(ProcessContext context) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testSimpleStateIdAnonymousDoFn() throws Exception {
+    DoFnSignature sig = DoFnSignatures.INSTANCE.getOrParseSignature(
+        new DoFn<KV<String, Integer>, Long>() {
+          @StateId("foo")
+          StateSpec<Object, ValueState<Integer>> bizzle = StateSpecs.value(VarIntCoder.of());
+
+          @ProcessElement
+          public void foo(ProcessContext context) {}
+        }.getClass());
+
+    assertThat(sig.stateDeclarations().size(), equalTo(1));
+    DoFnSignature.StateDeclaration decl = sig.stateDeclarations().get("foo");
+
+    assertThat(decl.id(), equalTo("foo"));
+    assertThat(decl.field().getName(), equalTo("bizzle"));
+    assertThat(
+        decl.stateType(),
+        Matchers.<TypeDescriptor<?>>equalTo(new TypeDescriptor<ValueState<Integer>>() {}));
+  }
+
+  @Test
+  public void testSimpleStateIdNamedDoFn() throws Exception {
+    // Test classes at the bottom of the file
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.signatureForDoFn(new DoFnForTestSimpleStateIdNamedDoFn());
+
+    assertThat(sig.stateDeclarations().size(), equalTo(1));
+    DoFnSignature.StateDeclaration decl = sig.stateDeclarations().get("foo");
+
+    assertThat(decl.id(), equalTo("foo"));
+    assertThat(
+        decl.field(), equalTo(DoFnForTestSimpleStateIdNamedDoFn.class.getDeclaredField("bizzle")));
+    assertThat(
+        decl.stateType(),
+        Matchers.<TypeDescriptor<?>>equalTo(new TypeDescriptor<ValueState<Integer>>() {}));
+  }
+
+  @Test
+  public void testGenericStatefulDoFn() throws Exception {
+    // Test classes at the bottom of the file
+    DoFn<KV<String, Integer>, Long> myDoFn = new DoFnForTestGenericStatefulDoFn<Integer>(){};
+
+    DoFnSignature sig = DoFnSignatures.INSTANCE.signatureForDoFn(myDoFn);
+
+    assertThat(sig.stateDeclarations().size(), equalTo(1));
+    DoFnSignature.StateDeclaration decl = sig.stateDeclarations().get("foo");
+
+    assertThat(decl.id(), equalTo("foo"));
+    assertThat(
+        decl.field(), equalTo(DoFnForTestGenericStatefulDoFn.class.getDeclaredField("bizzle")));
+    assertThat(
+        decl.stateType(),
+        Matchers.<TypeDescriptor<?>>equalTo(new TypeDescriptor<ValueState<Integer>>() {}));
+  }
+
+  private static class DoFnForTestSimpleStateIdNamedDoFn extends DoFn<KV<String, Integer>, Long> {
+    @StateId("foo")
+    StateSpec<Object, ValueState<Integer>> bizzle = StateSpecs.value(VarIntCoder.of());
+
+    @ProcessElement
+    public void foo(ProcessContext context) {}
+  }
+
+  private static class DoFnForTestGenericStatefulDoFn<T> extends DoFn<KV<String, T>, Long> {
+    // Note that in order to have a coder for T it will require initialization in the constructor,
+    // but that isn't important for this test
+    @StateId("foo")
+    StateSpec<Object, ValueState<T>> bizzle;
```

Comment:
Makes me think we should add some validation that all state spec fields are non-null at `ParDo` apply time. I would actually propose we require these fields be final, and check they are non-null when we `ParDo.apply()`. Also, they should probably be non-transient (so that they can be serialized properly). I think it would be OK for them to be static.